psi4/src/psi4/lib3index/dfhelper.cc

```diff
@@ -252,13 +258,13 @@ void DFHelper::prepare_sparsity() {
     // prep info vectors
     std::vector<double> shell_max_vals(pshells_ * pshells_, 0.0);
     std::vector<double> fun_max_vals(nao_ * nao_, 0.0);
-    schwarz_shell_mask_.reserve(pshells_ * pshells_);
-    schwarz_fun_mask_.reserve(nao_ * nao_);
-    symm_ignored_columns_.reserve(nao_);
-    symm_big_skips_.reserve(nao_ + 1);
-    symm_small_skips_.reserve(nao_);
-    small_skips_.reserve(nao_ + 1);
-    big_skips_.reserve(nao_ + 1);
+    schwarz_shell_mask_.resize(pshells_ * pshells_);
+    schwarz_fun_mask_.resize(nao_ * nao_);
+    symm_ignored_columns_.resize(nao_);
+    symm_big_skips_.resize(nao_ + 1);
+    symm_small_skips_.resize(nao_);
+    small_skips_.resize(nao_ + 1);
+    big_skips_.resize(nao_ + 1);
```

Comment:
The reason for reserve vs resize is want an exact allocation as resize will usually oversubscribe the array in case of new pushbacks. In addition resize will zero out the array which will hit performance as we assume uninitialized vectors. A limitation of `std::vector` which isn't exactly the correct container for this. @schiebermc any thoughts?

Reply:
Did something break ?
I essentially used `std::vector` as a glorified array.  In each case I knew the required size so I used `reserve` to allocate the necessary buffers.  While this gets me the memory I need and thereby increases `vector::capacity`, it does not affect `vector::size`.

Reply:
No, this is fine to me. Curious if there was another way to do this using the standard library if this is a MSVC issue.

Reply:
I guess I am not sure what is "incorrect", does this break on MSVC?
I would agree the use of `std::vector` here is unorthodox.
However, adding the fill constructor and `resize()` instead of using `reserve()` will degrade performance. Of course, the cost of these two operations are O(n), and it might be the case that this cost is trivial in the wake of the O(n^4) operations elsewhere. On the other hand, neither of these are currently threaded (to my knowledge). I would want to time the two versions before making the change. If it is a barrier to porting, then it may be a necessary sacrifice.

Reply:
At a glance, reserve() might result in undefined behavior, since using the [ ] operator to access an element beyond the *size* of the vector is undefined, even though in practice, if the underlying memory has been allocated with reserve() it will probably work, and not even show up as a memory leak in Valgrind. (since the destructor deallocates the *capacity* of the vector) Good rule of thumb, is that if you could replace all [ ] operators with .at() without getting exceptions, then [ ] is not relying on UB.

Reply:
@jturney ran the code through address and / or memory sanitizer at the last hackathon, and ended up changing a number of `reserve` -> `resize`.  That would be a good check to perform here.  Can you remember the details, Jet?

Reply:
The issue comes down to the fact that `reserve` creates space for the objects, without creating the objects (ints in this case), IIRC

Reply:
@andysim Yes, when a vector is enlarged via reserve(), internally it ends up calling malloc() (or equivalent) at some point, copies the contents of the original memory block into the new one, deallocates the old block and updates all relevant members of the vector object. (such as *capacity*)
Since *int* or *double* does not have any constructor, in this case relying on UB is stealthy. 
But if someone were to actually use some of the features offered by the vector object, very evil bugs could surface: for example if someone were to use the assignment operator to make a copy of the vector, then only *size* elements would be copied to the new vector, and any data between *size* and *reserve* would be missing. 
So IMO there are a ton of good reasons why elements beyond the *size* of a std::vector should never be accessed.

Reply:
Can we change all of the `std::vectors` in this file, but outside of this block (starting at line 388) to the following:
```C++
std::unique_ptr<double[]> array_name(new double[size_of_array]);
array_name_ptr = array_name.get();
```
We get the same effect in a slightly more canonical way. If these is too much we can patch it up after the fact.

Convo with @bennybp, @schiebermc sound reasonable to you?

Reply:
As @TiborGY already mentioned, it is undefined behaviour. MSVC gives invalid memory access errors, if the vectors are not properly resized.

@dgasmith, I will try your idea with `std::unique_ptr`.

Reply:
Sounds reasonable to me.